contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+                } else {
+                    _aggreateSide.set(order.price, _aggreateSide.get(order.price) + order.size);
+                }
+            });
+            // Construct array for our orderbook system
+            let __side: any[][];
+            if (_aggreateSide) {
+                __side = Array.from(_aggreateSide.entries()).map((_side_) => [
+                    (_side_[0].toFixed(4)),
+                    _side_[1],
+                ]);
+            } else {
+                // Return default values of 0
+                return [[0, 0]];
+            }
+            // Update our values for the orderbook
+            if (side === 'asks') {
+                setAsks(__side);
+            } else {
+                setBids(__side);
+            }
+            setLastSlotUpdated(ctx.slot);
+            // Check that we have books
+
+            let tobAsk: number;
+            let tobBid: number;
+
+            // Get top of books
+            if (side === 'asks') {
+                tobAsk = Number(__side[0][0]);
+                // @ts-ignore
+                tobBid = Number(bids[0][0]);
+            } else {
+                // @ts-ignore
+                tobAsk = Number(asks[0][0]);
+                tobBid = Number(__side[0][0]);
+            }
+            // Calculate spread
+            const spread: number = tobAsk - tobBid;
+            // Calculate spread percent
+            const spreadPercent: string = ((spread / tobBid) * 100).toFixed(2);
+            let _spreadString: string;
+            // Create our string for output into the orderbook object
+            if (spread === tobAsk) {
+                _spreadString = '∞';
+            } else {
+                _spreadString = `${spread.toFixed(2).toString()} (${spreadPercent}%)`;
+            }
+            setSpreadString(
+                (curSpreadString) => curSpreadString === _spreadString ? curSpreadString : _spreadString
+            );
+
+            setWsConnected((curConnected) => curConnected === false);
+        } catch (err) {
+            // console.error(err);
+            // TODO: Add in call to analytics / reporting
+        }
+    };
+
+    useEffect(() => {
+        if (isPassMarket) {
+            if (!bids) {
+                setBids(orderBookObject?.passBidsArray);
+            }
+            if (!asks) {
+                setAsks(orderBookObject?.passAsksArray);
+            }
+            if (!spreadString) {
+                setSpreadString(orderBookObject?.passSpreadString);
+            }
+        } else {
+            if (!bids) {
+                setBids(orderBookObject?.failBidsArray);
+            }
+            if (!asks) {
+                setAsks(orderBookObject?.failAsksArray);
+            }
+            if (!spreadString) {
+                setSpreadString(orderBookObject?.failSpreadString);
+            }
+        }
+    });
+
+    const listenOrderBook = async () => {
```

Comment:
I see that this has been moved here, excellent! However now we have BOTH markets available to us, so we can use the markets = [fail, pass] and let the foreach take care of the subscription.

Reply:
yeaahhh haha I have a different branch where I had all this right but in the openbook madness I stashed a bunch of things and switched branches and so this is an intermediary step of sorts. 

But yeah I think we might as well move this forward right now since I just need to switch the orderbook to read from here and put the state here as well.

Reply:
Okay I just implemented the updates so orderbook ask/bid side is being stored in context and read from their. No more order books local state it's in the context so we can use this down the road for updating user orders.